java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java

```diff
@@ -165,18 +166,40 @@ public void registerSSHMinion(String minionId, Optional<Long> proxyId,
     private void registerMinion(String minionId, boolean isSaltSSH,
                                 Optional<Long> saltSSHProxyId,
                                 Optional<String> activationKeyOverride) {
+
         // Match minions via their machine id
-        Optional<String> optMachineId = SALT_SERVICE.getMachineId(minionId);
-        if (!optMachineId.isPresent()) {
-            LOG.info("Cannot find machine id for minion: " + minionId);
-            return;
-        }
-        //FIXME: refactor this whole function so we don't have to call get
-        //in so many places.
-        String machineId = optMachineId.get();
-        MinionServer minionServer;
-        Optional<User> creator = MinionPendingRegistrationService.getCreator(minionId);
+        Opt.fold(SALT_SERVICE.getMachineId(minionId),
+            () -> {
+                LOG.info("Cannot find machine id for minion: " + minionId);
+                return false;
+            },
+            machineId -> {
+                Optional<User> creator = MinionPendingRegistrationService.getCreator(minionId);
+                if (checkIfMinionAlreadyRegistered(minionId, machineId, creator, isSaltSSH)) {
+                    return true;
+                }
+                if (duplicateMinionNamePresent(minionId)) {
+                    return false;
```

Comment:
If we abort the registration here, we might want to log an error.
Idea: or is this a case for a notification message?

Reply:
method `duplicateMinionNamePresent` already logs it with `error` severity. Maybe a notification message is not a bad idea. @ncounter any thoughts?

Reply:
@cavalheiro @mcalmer : good point!! We are already handling [here](https://github.com/uyuni-project/uyuni/blob/75ca4d31583956bff0bc24a72c491d8a90aa3fff/java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java#L1092) unexpected exceptions for an `OnboardingFailed` event. Maybe we can just improve/extend this `NotificationMessage` type to save a fixed output message (at the moment this type receives only a `minionId` string as a body context message).

Reply:
I totally agree this is out of this PR scope.
Maybe here you can just make sure that, for now, at least it throws [this exception](https://github.com/uyuni-project/uyuni/blob/75ca4d31583956bff0bc24a72c491d8a90aa3fff/java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java#L360) that is already handled and it creates a simple `NotificationMessage` about the onboarding failure with the `minionId`.
To have a better specific `NotificationMessage` I would go with a separate proper issue.

Reply:
This will require that I move the call to `duplicateMinionNamePresent` inide the `finalizeMinionRegistration` method so that I have access to the org information (required by this exception). But I think that will not be a problem.

Reply:
Yes, adding notification messages should be done extra.